File: source/geomagneticfield/deriv_legendre.c
double DerivLegendre(double n, double m, double x) {
  double d = 0;
  if (m == 0) {
    d = (-1) * AssLegendre(n, 1, x) / sqrt(1-pow(x, 2));
  } else {
    double d_a, d_b, d_c;
    d_a = sqrt(2 * factorial(n-m)/factorial(n+m));
    d_b = (-m)*x/(1-pow(x, 2))*pow(-1, m)*AssLegendre(n, m, x);
    if (m == n)
      d_c = 0;
    else
      d_c = pow(-1, m+1) / sqrt(1-pow(x, 2)) * AssLegendre(n, m+1, x);
    d = d_a * (d_b + d_c);
  }
  return d;
}

File: Matlab/GeomagneticField/derivLegendre.m
function d = derivLegendre(n, m, x)
% Calculate derivative of Schmidt seminormalized associated Legendre polynomial with respect to x

if m == 0
    d = -1 * assLegendre(n, 1, x) / sqrt(1 - x^2);
else
    d_a = sqrt(2 * factorial(n - m) / factorial(n + m));
    d_b = -m * x / (1 - x^2) * (-1)^m * assLegendre(n,m,x);
    
    if m == n
        d_c = 0;
    else
        d_c = (-1)^(m + 1) / sqrt(1 - x^2) * assLegendre(n, m+1, x);
    end
    
    d = d_a * (d_b + d_c);
end